Game/data/building_components/component_manager.gd

```diff
@@ -9,50 +9,69 @@ signal component_added(component: BuildingComponent)
+## List of array indices that are vacant in components_list.
+## Used when inserting into components_list.
+var _vacant_indices: Array[int]
 
 
 ## Initialize components by adding components to the BuildingEntity and the component list.
 func init_components_building(building: BuildingEntity) -> void:
-	var building_resource: BuildingResource = Buildings.get_by_id(building.id)
+	var building_resource: BuildingResource = Buildings.get_by_id(building.building_id)
 	for component_data: BuildingComponentData in building_resource.building_components:
-		var component: BuildingComponent = component_data.make_component(building)
-		add_component(component)
+		# add component to components_list
+		var unique_id: int
+		var component: BuildingComponent
+		if not _vacant_indices.is_empty():
+			unique_id = _vacant_indices.pop_back()
+			component = component_data.make_component(unique_id, building)
+			_components_list[unique_id] = component
+		else:
+			unique_id = _components_list.size()
+			component = component_data.make_component(unique_id, building)
+			_components_list.append(component)
+
+		# add component to components_by_type
+		if not _components_by_type.has(component.type):
+			_components_by_type[component.type] = []
+		_components_by_type[component.type].append(unique_id)
+
+		# add component to building's component list
 		building.components.append(component)
 
+		component_added.emit(component)
+		print("added component of type %s, component id %d" % [component.type, unique_id])
+		print("component type length is now %d" % _components_by_type[component.type].size())
+
 
 ## Remove an existing component with the ComponentManager.
 ## Returns true if it was removed, and false if it wasn't.
 func remove_component(component: BuildingComponent) -> bool:
-	var index_to_remove: int = -1
-	var components: Array = _components_list.get(component.type)
-	if not components:
+	var unique_id: int = component.unique_id
+	if unique_id >= _components_list.size():
+		assert(
+			false,
+			(
+				"Attempting to remove component with unique id %d bigger than components_list.size() %d"
+				% [unique_id, _components_list.size()]
+			)
+		)
 		return false
 
-	for index: int in range(components.size()):
-		var curr_component: BuildingComponent = components[index]
-		if curr_component == component: # compare by reference is desired behavior
-			index_to_remove = index
-
-	if index_to_remove != -1:
-		components.remove_at(index_to_remove)
-		component_removed.emit(component)
-		print("removed component of type %s" % [component.type])
-		print("components length is now %d" % _components_list[component.type].size())
-		return true
-	else:
-		return false
+	_components_list[unique_id] = null
+	_components_by_type[component.type].erase(unique_id)
+	_vacant_indices.append(unique_id)
+
+	component_removed.emit(component)
+	print("removed component of type %s, component id %d" % [component.type, unique_id])
```

Comment:
Change these to print_debug or remove them.

Reply:
done